Fix invalid instantiation and possibly unsafe accesses of array in class base_uint<BITS>

pavlosantoniou opened the pull request:
The implementation of base_uint<BITS>::operator++(int) and base_uint<BITS>::operator--(int) is now safer.
Array pn is accessed via index i after bounds checking has been performed on the index, rather than before.

The logic of the while loops has also been made more clear.

laanwj commented on src/arith_uint256.h:
```diff
@@ -174,7 +174,7 @@ class base_uint
         // prefix operator
         int i = 0;
-        while (++pn[i] == 0 && i < WIDTH-1)
+        while (i < WIDTH && ++pn[i] == 0)
```
As I see it, this code can cause an out-of-bounds access only in the case of WIDTH=0, which would be pointless.
Or am I missing something?
I do agree that this change makes the code somewhat more readable.

pavlosantoniou replied:
Yes, the out-of-bounds access occurs in the edge case of WIDTH == 0.
This template class can, although it probably won't, be instantiated with BITS < 32, something that would lead to WIDTH == 0. 
I was mainly interested in ensuring the out-of-bounds-check before array access regardless of such assumptions.
Making it easier to follow the code logic is an extra gain.

laanwj replied:
> This template class can, although it probably won't, be instantiated with BITS < 32

Indeed - as the code throughout the entire class is implemented, it can only be instantiated for a positive multiple of 32. Might make sense to add a compile-time assertion for that.

pavlosantoniou replied:
Do you think adding:
`static_assert(BITS/32 > 0 && BITS%32 == 0, "Template parameter BITS must be a positive multiple of 32."");`
in the class constructor would be fine?

laanwj replied:
Yes, that'd be nice!

pavlosantoniou replied:
Is there something more I can do for this PR?

pavlosantoniou replied:
I have added the assertions, is there something more I can do for this PR?

pavlosantoniou committed:
Fix instantiation and array accesses in class base_uint<BITS>

The implementation of base_uint::operator++(int) and base_uint::operator--(int) is now safer.
Array pn is accessed via index i after bounds checking has been performed on the index, rather than before.
The logic of the while loops has also been made more clear.

A compile time assertion has been added in the class constructors to ensure that BITS is a positive multiple of 32.

pavlosantoniou commented:
Reopened with addition of static_assert for BITS value.

sipa commented:
utACK
